File: packages/core/src/scanner/ScanCache.ts
import { readFileSync, writeFileSync, existsSync, mkdirSync } from 'node:fs';
import { dirname } from 'node:path';
import type { ModelFinding } from '../types/Finding.js';

interface CacheState {
  entries: Record<string, CacheEntry>;
}

interface CacheEntry {
  cachedAtUtc: string;
  findings: ModelFindingCache[];
}

interface ModelFindingCache {
  techniqueId?: string;
  chunkId: string;
  file: string;
  startLine: number;
  endLine: number;
  severity: string;
  confidence: number | null;
  observation: string;
  evidence: string;
  reasoning: string | null;
  unknownMitigations: string[];
  modelName: string;
}

export class ScanCache {
  private state: CacheState;
  private pendingWrites = 0;
  private lastFlushMs = Date.now();

  constructor(private readonly cacheFile: string) {
    this.state = this.load();
  }

  get(key: string): ModelFinding[] | null {
    const entry = this.state.entries[key];
    if (!entry) return null;
    return entry.findings.map((finding) => toModelFinding(finding));
  }

  put(key: string, findings: ModelFinding[]): void {
    this.state.entries[key] = {
      cachedAtUtc: new Date().toISOString(),
      findings: findings.map((finding) => fromModelFinding(finding)),
    };
    this.pendingWrites += 1;
  }

  flush(): void {
    this.flushLocked();
  }

  flushMaybe(): void {
    if (this.pendingWrites === 0) return;
    const now = Date.now();
    const shouldFlush =
      this.pendingWrites >= 20 || now - this.lastFlushMs >= 5000;
    if (shouldFlush) {
      this.flushLocked();
    }
  }

  private flushLocked(): void {
    try {
      mkdirSync(dirname(this.cacheFile), { recursive: true });
      writeFileSync(this.cacheFile, JSON.stringify(this.state), 'utf8');
      this.pendingWrites = 0;
      this.lastFlushMs = Date.now();
    } catch {
      // best-effort cache
    }
  }

  private load(): CacheState {
    try {
      if (!existsSync(this.cacheFile)) {
        return { entries: {} };
      }
      const raw = readFileSync(this.cacheFile, 'utf8');
      const parsed = JSON.parse(raw) as CacheState;
      if (!parsed.entries) {
        return { entries: {} };
      }
      return parsed;
    } catch {
      return { entries: {} };
    }
  }
}

function toModelFinding(cache: ModelFindingCache): ModelFinding {
  return {
    techniqueId: cache.techniqueId,
    chunkId: cache.chunkId,
    file: cache.file,
    startLine: cache.startLine,
    endLine: cache.endLine,
    severity: cache.severity,
    confidence: cache.confidence,
    observation: cache.observation,
    evidence: cache.evidence,
    reasoning: cache.reasoning,
    unknownMitigations: cache.unknownMitigations,
    modelName: cache.modelName,
  };
}

function fromModelFinding(finding: ModelFinding): ModelFindingCache {
  return {
    techniqueId: finding.techniqueId,
    chunkId: finding.chunkId,
    file: finding.file,
    startLine: finding.startLine,
    endLine: finding.endLine,
    severity: finding.severity,
    confidence: finding.confidence,
    observation: finding.observation,
    evidence: finding.evidence,
    reasoning: finding.reasoning,
    unknownMitigations: finding.unknownMitigations,
    modelName: finding.modelName,
  };
}

File: packages/intellij-plugin/src/main/kotlin/ai/astha/scantheplanet/idea/scanner/ScanCache.kt
package ai.astha.scantheplanet.idea.scanner

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.KotlinModule
import com.fasterxml.jackson.module.kotlin.readValue
import ai.astha.scantheplanet.idea.scanner.model.ModelFinding
import java.nio.file.Files
import java.nio.file.Path
import java.time.ZonedDateTime

class ScanCache(private val cacheFile: Path) {
    private val mapper = ObjectMapper().registerModule(KotlinModule.Builder().build())
    private val lock = Any()
    private var state: CacheState = load()
    private var pendingWrites = 0
    private var lastFlushNanos = System.nanoTime()

    fun get(key: String): List<ModelFinding>? {
        synchronized(lock) {
            val entry = state.entries[key] ?: return null
            return entry.findings.map { it.toModelFinding() }
        }
    }

    fun put(key: String, findings: List<ModelFinding>) {
        synchronized(lock) {
            state.entries[key] = CacheEntry(
                cachedAtUtc = ZonedDateTime.now(java.time.ZoneOffset.UTC).toString(),
                findings = findings.map { ModelFindingCache.from(it) }
            )
            pendingWrites += 1
        }
    }

    fun flush() {
        synchronized(lock) {
            flushLocked(System.nanoTime())
        }
    }

    fun flushMaybe() {
        synchronized(lock) {
            if (pendingWrites == 0) return
            val now = System.nanoTime()
            val shouldFlush = pendingWrites >= 20 || (now - lastFlushNanos) >= 5_000_000_000L
            if (shouldFlush) {
                flushLocked(now)
            }
        }
    }

    private fun flushLocked(nowNanos: Long) {
        try {
            Files.createDirectories(cacheFile.parent)
            mapper.writeValue(cacheFile.toFile(), state)
            pendingWrites = 0
            lastFlushNanos = nowNanos
        } catch (_: Exception) {
            // best-effort cache
        }
    }

    private fun load(): CacheState {
        return try {
            if (Files.exists(cacheFile)) {
                mapper.readValue(cacheFile.toFile())
            } else {
                CacheState()
            }
        } catch (_: Exception) {
            CacheState()
        }
    }

    data class CacheState(
        val entries: MutableMap<String, CacheEntry> = mutableMapOf()
    )

    data class CacheEntry(
        val cachedAtUtc: String,
        val findings: List<ModelFindingCache>
    )

    data class ModelFindingCache(
        val techniqueId: String? = null,
        val chunkId: String,
        val file: String,
        val startLine: Int,
        val endLine: Int,
        val severity: String,
        val confidence: Double?,
        val observation: String,
        val evidence: String,
        val reasoning: String?,
        val unknownMitigations: List<String>,
        val modelName: String
    ) {
        fun toModelFinding(): ModelFinding {
            return ModelFinding(
                techniqueId = techniqueId,
                chunkId = chunkId,
                file = file,
                startLine = startLine,
                endLine = endLine,
                severity = severity,
                confidence = confidence,
                observation = observation,
                evidence = evidence,
                reasoning = reasoning,
                unknownMitigations = unknownMitigations,
                modelName = modelName
            )
        }

        companion object {
            fun from(finding: ModelFinding): ModelFindingCache {
                return ModelFindingCache(
                    techniqueId = finding.techniqueId,
                    chunkId = finding.chunkId,
                    file = finding.file,
                    startLine = finding.startLine,
                    endLine = finding.endLine,
                    severity = finding.severity,
                    confidence = finding.confidence,
                    observation = finding.observation,
                    evidence = finding.evidence,
                    reasoning = finding.reasoning,
                    unknownMitigations = finding.unknownMitigations,
                    modelName = finding.modelName
                )
            }
        }
    }
}
